lib/src/main/java/io/ably/lib/realtime/ChannelBase.java

```diff
@@ -843,6 +843,10 @@ private void onMessage(final ProtocolMessage protocolMessage) {
             if(msg.connectionId == null) msg.connectionId = protocolMessage.connectionId;
             if(msg.timestamp == 0) msg.timestamp = protocolMessage.timestamp;
             if(msg.id == null) msg.id = protocolMessage.id + ':' + i;
+            // (TM2p)
+            if(msg.version == null) msg.version = String.format("%s:%03d", protocolMessage.channelSerial, i);
+            // (TM2k)
+            if(msg.serial == null) msg.serial = msg.version;
```

Comment:
Also needed here is `TM2o` for the `createdAt` timestamp (needs to be specified if not provided by the server)

Reply:
@AndyTWF not sure why this was closed? 
We do the same in case of swift -> https://github.com/ably/ably-cocoa/blob/9f1cbd92c0456ccca486a6efedfceb572a0d7bb2/Source/ARTJsonLikeEncoder.m#L289

Reply:
Ah sorry, I'd opened a separate thing about the `msg.action` check for `createdAt` - but yes, the same applies to the `serial field`

Reply:
It makes no sense to me, especially in java, there NPE is most common reason of app crash, but I fixed this everywhere